### ISSUE TITLE
I found a bug and improved it with a huge performance boost

### ISSUE THREAD

**Issue by 18112330636** [XinzeLee/polygon-yolov5#20]
The core idea of this model is to regress the deviation of the four vertices of the quadrilateral to the center point. To simplify the problem, sort the vertices of the quad such that y3, y4 >= y1, y2; x1 <= x2; x4 <= x3. But there is a flaw. If the diagonal of the quadrilateral is parallel to the x-axis, then y2=y3, then this method is ambiguous, and the order of the left and right vertices on the diagonal is undefined. This ambiguity results in greatly reduced performance of the model. Moreover, three for loops are written in the method of sorting vertices, resulting in very slow training. Before the improvement, my data required training for 1000epoch to reach the 0.5 threshold of 94map. After the improvement, my data only needs to train 120epoch to reach the 0.5 threshold of 96map. Also, the pre-improved model cannot be scaled and rotated. After the improvement, the degree of polygon_random_perspective is changed to 90, the scale is changed to 0.5, and the model can still reach 96map at about 150epoch. Below is my code。I want to change jobs in Shanghai, China， Interested can contact me。 28 years old with three years of CV experience。number 18112330636

def order_corners(boxes):
    """
        Return sorted corners for loss.py::class Polygon_ComputeLoss::build_targets
        Sorted corners have the following restrictions:
                                y3, y4 >= y1, y2; x1 <= x2; x4 <= x3
    """
    if boxes.shape[0] == 0:
        return torch.empty(0, 8, device=boxes.device)

    boxes = boxes.view(-1, 4, 2)
    x = boxes[..., 0]
    y = boxes[..., 1]
    y_sorted, y_indices = torch.sort(y) # sort y
    idx = torch.arange(0, y.shape[0], dtype=torch.long, device=boxes.device)
    complete_idx = idx[:, None].repeat(1, 4)
    x_sorted = x[complete_idx, y_indices]
    x_sorted[:, :2], x_bottom_indices = torch.sort(x_sorted[:, :2])
    x_sorted[:, 2:4], x_top_indices = torch.sort(x_sorted[:, 2:4], descending=True)
    y_sorted[idx, :2] = y_sorted[idx, :2][complete_idx[:, :2], x_bottom_indices]
    y_sorted[idx, 2:4] = y_sorted[idx, 2:4][complete_idx[:, 2:4], x_top_indices]
    special = (y_sorted[:, 1] == y_sorted[:, 2]) & (x_sorted[:, 1] > x_sorted[:, 2])
    if idx[special].shape[0]  != 0:
        x_sorted_1 = x_sorted[idx[special], 1].clone()
        x_sorted[idx[special], 1] = x_sorted[idx[special], 2]
        x_sorted[idx[special], 2] = x_sorted_1
    return torch.stack((x_sorted, y_sorted), dim=2).view(-1, 8).contiguous()

**Comment by XinzeLee** [comment]
Can you kindly paste your comparison results here? I will take a look and modify the code if necessary. Thanks!

**Comment by XinzeLee** [comment]
I have checked your code, and I found that indeed you have modified the for loop to matrix slicing version. This feature is great and I will update the code. For the diagnal parallel problem, does it have a significant impact on the performance? Can you show me your training process and results before and after the modification? 
Thanks! Will mention your contribution on this repo!

**Comment by 18112330636** [comment]
[results.txt](https://github.com/XinzeLee/PolygonObjectDetection/files/8520392/results.txt)
[results.txt](https://github.com/XinzeLee/PolygonObjectDetection/files/8520398/results.txt)
Shàngmiàn shì wǒ de liǎng cì xùnliàn jiéguǒ, dì yī gè wénjiàn shì yuánlái de order_corners xùnliàn jiéguǒ, dì èr gè shì yòng wǒ gǎijìn zhīhòu de xùnliàn jiéguǒ, kěyǐ hěn míngxiǎn de kàn chū wǒ de gǎijìn tíshēngle móxíng dì xìngnéng. Bìngqiě dì yī gè wénjiàn de xùnliàn jiéguǒ shì duì diǎn zuòle páixù de qíngkuàng xià de, rúguǒ bù zuò páixù, shùn shízhēn hé nì shízhēn de biāozhù dōu yǒu, nàme dāng x zhóu hé sìbiānxíng duì jiǎo xiàn píngxíng de shíhòu de x2 hé x3 de bù quèdìng xìng huì gèng duō. Wǒ zhèngzài fù xiàn zhège jiéguǒ. Tóng lǐ, rúguǒ polygon_random_perspective de jiǎodù gǎi dà yīdiǎn, zhè zhǒng chōngtú huì biàn duō, móxíng xìngnéng huì jíjù xiàjiàng. Dì èr gè wénjiàn de xùnliàn jiéguǒ wǒ de polygon_random_perspective de jiǎodù shèzhì wèi 90,scale shèzhì wèi 0.5, Zhè huì ràng chǎngjǐng gèngjiā fùzá huà, dànshì réngrán néng hěn kuài shōuliǎn. Shāo hòu wǒ huì shàngchuán zuì yuánshǐ de duìbǐ jiéguǒ
展开
315 / 5,000
翻译结果
The above are my two training results. The first file is the original order_corners training result, and the second is the training result after my improvement. It can be clearly seen that my improvement has improved the performance of the model. And the training result of the first file is when the points are sorted. If the sorting is not done, there are clockwise and counterclockwise labels, then when the x-axis and the diagonal of the quadrilateral are parallel, x2 and x3 will be more uncertain. I am reproducing this result. In the same way, if the angle of polygon_random_perspective is changed a little larger, there will be more such conflicts, and the model performance will drop sharply. The training result of the second file I have polygon_random_perspective angle set to 90 and scale set to 0.5, which makes the scene more complicated, but still converges quickly. I will upload the most original comparison results later

**Comment by 18112330636** [comment]
I still have a little idea of ​​improving this model. When the data needs to be clipped after polygon_random_perspective, the coordinates of the target will exceed the image, and this model does not do it, but when the center of the target is in the image. , but this is clearly problematic. This problem can be divided into three cases. The target obtained after the target overflow image is cropped may be a triangle, a quadrilateral, and a pentagon. I think it's possible to do some filtering and then clip the coordinates of the quad to the inside of the image. The accuracy should be improved at this time. I can't think of a good solution for the pentagon.

**Comment by XinzeLee** [comment]
Yes, I made the modifications to the def order_corners accordingly. 

By the way, you concerns regarding the polygon_random_perspective and mosaic also occurred to me before, but hard to find solutions. If you have further improvements, please let me know. Thanks!
